### PR TITLE
fix: signal disconnect loop on server start

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

this fixes the disconnect loop when the server has just started by waiting for the connection to open before subsribing to events.

also includes other improvements:
- rename types to interfaces
- allow to input single receiver for signal
- use discovery api to check for upgrade path instead hard coded one

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by drodil** [comment]
@Rugvip if you have time, please check this before next release, thanks!

File: .changeset/gorgeous-ways-applaud.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+'@backstage/plugin-signals-backend': patch
+'@backstage/plugin-signals-react': patch
+'@backstage/plugin-signals-node': patch
+'@backstage/plugin-signals': patch
+---
+
+Fix to disconnect loop on server start
```
Comment (Rugvip) [inline .changeset/gorgeous-ways-applaud.md:8]:
```suggestion
Fix disconnect loop on server start
```

I'm assuming you meant something like this 😅

File: plugins/signals-backend/src/service/router.ts [hunk 2]
```diff
@@ -66,9 +70,9 @@ export async function createRouter(
     }
 
     subscribedToUpgradeRequests = true;
+    const apiUrl = await discovery.getBaseUrl('signals');
     server.on('upgrade', async (request, socket, head) => {
-      // TODO: Find a way to make this more generic
-      if (request.url !== '/api/signals') {
+      if (!request.url || !apiUrl.endsWith(request.url)) {
```
Comment (Rugvip) [inline plugins/signals-backend/src/service/router.ts:75]: 👍

File: plugins/signals-node/api-report.md [hunk 3]
```diff
@@ -16,15 +16,15 @@ export class DefaultSignalService implements SignalService {
 
 // @public (undocumented)
 export type SignalPayload = {
-  recipients: string[] | null;
+  receivers: string[] | string | null;
```
Comment (Rugvip) [inline plugins/signals-node/api-report.md:19]: Hmm, not sure I like receivers more but not gonna stress it for now

**Review by Rugvip — APPROVED** [review]
Looks good! 👍 🎉 

Gonna :shipit: with the lil changeset tweak

**Comment by drodil** [comment]
Some checks not passing anymore but seems like not related to this change

File: plugins/signals-node/api-report.md (hunk 3 above)
Comment (Rugvip) [reply plugins/signals-node/api-report.md:19]: XD

Comment (Rugvip) [reply plugins/signals-node/api-report.md:19]: Hmm, should we switch it back and change the BEP? Feels like that's where we're going anyway

**Review by Rugvip — APPROVED** [review]
👌 :shipit:

**Comment by drodil** [comment]
@Rugvip what to do with the failing builds?

**Pull request merged**